src/Aws/Lambda/Runtime.hs

```diff
@@ -29,65 +20,103 @@ import qualified Aws.Lambda.Runtime.Environment as Environment
+import Control.Exception.Safe.Checked
+import qualified Control.Exception.Safe.Checked as Checked
+import Control.Monad (forever)
+import Data.Aeson
+import qualified Data.ByteString.Lazy as Lazy
+import Data.IORef
+import qualified Network.HTTP.Client as Http
+import System.IO (hFlush, stderr, stdout)
 
--- | Runs the user @haskell_lambda@ executable and posts back the
--- results. This is called from the layer's @main@ function.
-runLambda :: forall context. IO context -> Runtime.RunCallback context -> IO ()
-runLambda initializeCustomContext callback = do
+runLambda ::
+  ( FromJSON event,
+    Runtime.ToLambdaResponseBody error,
+    Runtime.ToLambdaResponseBody result
+  ) =>
+  IO context ->
+  (event -> Context.Context context -> IO (Either error result)) ->
+  IO ()
+runLambda initializer userCode =
+  runtime initializer (wrapStandaloneLambda userCode)
+
+wrapStandaloneLambda ::
+  ( FromJSON event,
+    Runtime.ToLambdaResponseBody error,
+    Runtime.ToLambdaResponseBody result
+  ) =>
+  (event -> Context.Context context -> IO (Either error result)) ->
+  Runtime.LambdaOptions context ->
+  IO (Either Runtime.LambdaError Runtime.LambdaResult)
+wrapStandaloneLambda userCode Runtime.LambdaOptions {..} =
+  case eitherDecode eventObject of
+    Left err ->
+      pure . Left . Runtime.StandaloneLambdaError $ Runtime.toStandaloneLambdaResponse err
+    Right event -> do
+      result <- userCode event contextObject
+      case result of
+        Left err ->
+          pure . Left . Runtime.StandaloneLambdaError $ Runtime.toStandaloneLambdaResponse err
+        Right ok ->
+          pure . Right . Runtime.StandaloneLambdaResult $ Runtime.toStandaloneLambdaResponse ok
+
+runtime :: forall context. IO context -> Runtime.RunCallback context -> IO ()
+runtime initializeCustomContext callback = do
   manager <- Http.newManager httpManagerSettings
   customContext <- initializeCustomContext
   customContextRef <- newIORef customContext
-  context   <- Context.initialize @context customContextRef `catch` errorParsing `catch` variableNotSet
+  context <- Context.initialize @context customContextRef `catch` errorParsing `catch` variableNotSet
   forever $ do
     lambdaApi <- Environment.apiEndpoint `catch` variableNotSet
-    event     <- ApiInfo.fetchEvent manager lambdaApi `catch` errorParsing
-
+    event <- ApiInfo.fetchEvent manager lambdaApi `catch` errorParsing
     -- Purposefully shadowing to prevent using the initial "empty" context
     context <- Context.setEventData context event
-
-    (((invokeAndRun callback manager lambdaApi event context
-      `Checked.catch` \err -> Publish.parsingError err lambdaApi context manager)
-      `Checked.catch` \err -> Publish.invocationError err lambdaApi context manager)
-      `Checked.catch` \(err :: Error.EnvironmentVariableNotSet) -> Publish.runtimeInitError err lambdaApi context manager)
+    ( ( ( invokeAndRun callback manager lambdaApi event context
```

Comment:
I am not sure if you are forced to use `safe-exceptions-checked` by one of your dependencies, but if you are not, `catches` in `Control.Exception` may be tidier here: https://hackage.haskell.org/package/base-4.12.0.0/docs/Control-Exception.html#v:catches